src/core.jl

```diff
@@ -122,20 +131,26 @@ isapprox(n::Node, f::Node) = unbox(n) ≈ unbox(f)
 zero(n::Node) = zero(unbox(n))
 one(n::Node) = one(unbox(n))
 
+# Let the user get the `size` and `length` of `Node`s.
+Base.size(x::Node, dims...) = size(unbox(x), dims...)
+Base.length(x::Node) = length(unbox(x))
+
 # Leafs do nothing, Branches compute their own sensitivities and update others.
 @inline propagate(y::Leaf, rvs_tape::Tape) = nothing
 function propagate(y::Branch, rvs_tape::Tape)
-    tape = Nabla.tape(rvs_tape)
-    ȳ, f = tape[pos(y)], getfield(y, :f)
+    ȳ = rvs_tape[y]  # the gradient we are going to propagate through the operation in y
+    d_tape = Nabla.tape(rvs_tape)  # strips off the Tape abstration leaving a plain Vector
+    f = getfield(y, :f)
     args = getfield(y, :args)
     kwargs = getfield(y, :kwargs)
-    xs, xids = map(unbox, args), map(pos, args)
-    p = preprocess(f, unbox(y), ȳ, xs...)
+    xs = map(unbox, args)
+    xids = map(pos, args)
+    p = preprocess(f, y, ȳ, args...)  # inlining CSE will avoid unboxing twice.
```

Comment:
should check before calling `preprocess` if `ȳ` is a `AbstractZero` and if so don't bother calling preprocess etc and just propagate the AbstractZero.

Reply:
This can be a follow up PR